### ISSUE TITLE
Consider simplifying OpenTelemetryAutoConfiguration

### ISSUE THREAD

**Issue by vpavic** [spring-projects/spring-boot#36248]
`OpenTelemetryAutoConfiguration` in its current shape contributes quite a few beans to the application context. In reality, most of those beans aren't likely be used anywhere (outside of contributing to `OpenTelemetry` bean) and appear to have been registered as beans just for the sake of being overridable (as they're `@ConditionalOnMissingBean`). However, the same effect (with lesser impact to application context) could have been achieved by having `OpenTelemetry` `@Bean` method accepting `ObjectProvider` arguments for different OTel components it depends on. Consequently, if I'm taking full control and configuring `OpenTelemetry` bean on my own, I'll still have all those OTel-related beans contributed to the application context.

So my proposal is to look into minimizing the number of OTel-related beans contributed to the application context, ideally reducing it to just `OpenTelemetry` bean.

More generally, the Actuator appears to me to be getting more and more bulky with time, and I'd like to see its impact on application context to be minimized, where possible.

**Comment by wilkinsona** [comment]
Duplicates #34023.

**Comment by mhalbritter** [comment]
After talking about that, we decided to reopen the issue.

This issue (creating beans which are overridable which may then not be used) is a more general problem in most of our auto-configurations. Tackling this in a more general way may require effort in the Spring Framework. Before starting work on that, we should benchmark the impact of unused beans on the application context to assess if this is worth fixing.

**Comment by vpavic** [comment]
Thanks for taking another look at this and breaking #34023 into several smaller, more focused, issues - I was about to suggest that.

> This issue (creating beans which are overridable which may then not be used) is a more general problem in most of our auto-configurations. Tackling this in a more general way may require effort in the Spring Framework.

Maybe I'm missing something, but what's the point of having so many OpenTelemetry components registered as beans in the first place? In order to offer a possibility of customizing some bits of the configuration by consuming user-contributed beans, it isn't really necessary to have those same components registered as beans by auto-configuration.

**Comment by mhalbritter** [comment]
Most (if not all) of the Spring Boot auto-configurations work in that way. They offer beans which are `@ConditionalOnMissingBean` so that users can override them. That's why we did it this way for OpenTelemetry, too. Of course it's technically possible to achieve that with `ObjectProvider`, but before refactoring the whole auto-configuration, we want to take a step back and see if this problem can be solved in a more general way.

**Comment by joaquinjsb** [comment]
Hello,
I would like to add also another issue regarding the autoconfiguration, as of now, it doesn't seem to be easy to customize the different exporters to provide e.g. sslbundles, I had to implement all the beans myself

for example:

```java
@Bean
    OtlpGrpcSpanExporter otlpGrpcSpanExporter(OtlpTracingProperties properties,
                                              OtlpTracingConnectionDetails connectionDetails,
                                              SslBundles sslBundles) {
        final SslBundle sslBundle = sslBundles.getBundle("observability");
        final SSLContext sslContext = sslBundle.createSslContext();
        final X509TrustManager trustManager = (X509TrustManager) sslBundle.getManagers().getTrustManagers()[0];
        OtlpGrpcSpanExporterBuilder builder = OtlpGrpcSpanExporter.builder()
                .setEndpoint(connectionDetails.getUrl(Transport.GRPC))
                .setTimeout(properties.getTimeout())
                .setConnectTimeout(properties.getConnectTimeout())
                .setSslContext(sslContext, trustManager)
                .setCompression(properties.getCompression().name().toLowerCase(Locale.ROOT));
        properties.getHeaders().forEach(builder::addHeader);
        return builder.build();
    }

```

also for metrics, there is no grpc transport on the configuration currently.

Would it make sense to add a PR to add sslbundle support or it's better to just rework the whole picture?

thanks

**Comment by wilkinsona** [comment]
@joaquinjsb I don't think that's really related to this issue as what you're proposing is not a simplification, it's a new feature. Please open a separate issue. When you do so, please expand upon what you mean by "implement all the beans myself". As far as I can tell (and as far as you have shown) you only need to define a single bean to apply SSL to the application's `OtlpGrpcSpanExporter`.

**Comment by mhalbritter** [comment]
@joaquinjsb https://github.com/spring-projects/spring-boot/pull/44900 gives you customizers to apply the ssl bundle.

**Comment by wilkinsona** [comment]
We discussed this today and we're skeptical that reducing the number of beans related to OpenTelemetry will make a noticeable difference. Those beans aren't only used at startup, they're an important part of the OpenTelemetry infrastructure that is needed at runtime. Removing such a bean will only reduce by one the number of entries in a map in the bean factory and not much more. That minimal benefit doesn't seem worth it, particularly when it would be a breaking change as application may be relying on the presence of those beans for various reasons.

Separate to this, we do wonder if some sort of ephemeral bean support in Framework may be useful. Before proceeding with that, analysing a typical Boot app and identifying the number of beans that are only used at startup and then never needed again would be a good first step. We'd need to identify a noticeable benefit in terms of memory footprint and dependency resolution performance to take things further. @bclozel will investigate and open a Framework issue if the benefits look like being worthwhile.
